### PR TITLE
Refactor cljam.io.util.bin

### PR DESCRIPTION
**Pull request by alumi**
#### Summary
Refactored `cljam.io.util.bin`

#### Changes
- Rearrange functions
- Add docstrings
- Add tests

#### Tests

- `lein check` ✅ 
- `lein test :all` ✅ 
- `lein cloverage` ✅
- `lein eastwood` ✅
- `lein cljfmt check` ✅

### CODE REVIEW

**Comment by alumi** [comment]
> Just to confirm, memoization for reg->bins was dropped, but is it intended?
> If so (probably so), could you elaborate a little more on the reason just for future reference? It messed up the performance improvement by unboxed function calls? You were concerned about its memory consumption? Or for any other reason?

Ah, sorry I forgot to mention that 🙇 
The changes to `reg->bins` are intentional.
- I fixed it to use `transduce` to speed it up by 1.5x-2x.
- I decided to remove the `memoize` because it's not repeatedly called and it likely to have a high cache miss rate (not well tested). `reg->bins*` was fast (~10 μs) enough by itself compared to actual file I/O ops. As you said, this will benefit reducing CPU overhead and memory leak.
- I made it public to balance with `reg->bin`.

File: src/cljam/io/util/bin.clj [hunk 1]
```diff
@@ -1,83 +1,96 @@
...
             (+ (first-bin-of-level level) beg-bins)
             (recur (dec level))))
         0))))
+
+(defprotocol IBinningIndex
+  (get-chunks [this ref-idx bins])
+  (get-min-offset [this ref-idx beg])
+  (get-min-shift [this])
+  (get-depth [this])
+  (get-chr-names  [this]))
```
[the remark this answers is not in the document]
Comment (alumi) [reply src/cljam/io/util/bin.clj]: [Fixed](https://github.com/chrovis/cljam/compare/e2f153b62559cdaf85f0371f9aff982837b29a3c..aa624401f1e809c5c16586d413a008b414428244). Thanks!

**Comment by athos** [comment]
Thank you for describing your decisions behind the refinements! They all sound reasonable to me 🙆‍♂️

> I decided to remove the memoize because it's not repeatedly called and it likely to have a high cache miss rate (not well tested).

Yeah, I can imagine that. The `(beg, end)` space in general might be too sparse for memoization to work effectively (which depends heavily on how `reg->bins` is used, though). But even without memoization, users can memoize the function on their own if they really need.

**Comment by alumi** [comment]
Thank you for reviewing! 😸  @niyarin @athos

**Pull request merged**